File: golang/rknight.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

type Move struct {
	row int
	col int
}

type KnightSolver struct {
	size int
	solved bool
	board [][]int
	path []Move
}

// Move methods
func (m *Move) display() string {
	return fmt.Sprintf("%c%d", rune(65+m.row), m.col+1)
}

// KnightSolver methods
func (s *KnightSolver) init(size int) {
	fmt.Println("Initializing board. Size=", size)
	s.size = size
	s.solved = false
	s.board = make([][]int, size)
	for k := 0; k < size; k++ {
		s.board[k] = make([]int, size)
	}
	s.path = make([]Move, size*size)
}

func (s *KnightSolver) move(row int, col int, moveNumber int) bool {
	if s.solved {
		return true
	}
	if row < 0 || row >= s.size || col < 0 || col >= s.size {
		return false
	}
	if s.board[row][col] > 0 {
		return false
	}

	// This is a valid move, update everything and move on
	s.board[row][col] = moveNumber
	s.path[moveNumber-1].row = row
	s.path[moveNumber-1].col = col
	if moveNumber == s.size*s.size {
		s.solved = true
		return true
	}
	s.move(row+1, col+2, moveNumber+1)
	s.move(row+2, col+1, moveNumber+1)
	s.move(row+1, col-2, moveNumber+1)
	s.move(row+2, col-1, moveNumber+1)
	s.move(row-1, col-2, moveNumber+1)
	s.move(row-2, col-1, moveNumber+1)
	s.move(row-1, col+2, moveNumber+1)
	s.move(row-2, col+1, moveNumber+1)
	s.board[row][col] = 0
	return s.solved
}

func main() {
	size := 7
	startCol := 0
	startRow := 0

	// Parse cmd line args
	nArgs := len(os.Args)
	if nArgs > 1 {
		size, _ = strconv.Atoi(os.Args[1])
		if nArgs > 2 {
			startCol, _ = strconv.Atoi(os.Args[2])
			if nArgs > 3 {
				startRow, _ = strconv.Atoi(os.Args[3])
			}
		}
	}
	startPos := Move{row: startCol, col: startRow}
	fmt.Println("Starting solver from pos", startPos.display())

	solver := KnightSolver{}
	solver.init(size)
	found := solver.move(startCol, startRow, 1)
	if found {
		fmt.Println("Found solution")
		for k := 0; k < size*size; k++ {
			fmt.Println(solver.path[k].display())
		}
	} else {
		fmt.Println("Unable to find solution")
	}
}

File: python/rknight.py
import argparse
import time
import sys

if sys.version_info >= (3, 3):
    clock = time.perf_counter
else:
    clock = time.clock


class KnightSolver:
    def __init__(self, size):
        self.size = size
        self.size2 = size*size
        row = [0 for k in range(size)]
        self.board = [list(row) for k in range(size)]
        self.path = [(0, 0) for k in range(size*size)]
        self.solved = False

    def move(self, row, col, move_number):
        if self.solved:
            return True
        if row < 0 or row >= self.size or col < 0 or col >= self.size:
            return False
        if self.board[row][col] > 0:
            return False

        # The move is valid. Update everything and move on
        self.board[row][col] = move_number
        self.path[move_number-1] = (row, col)
        if move_number == self.size2:
            self.solved = True
            return True
        self.move(row+1, col+2, move_number+1)
        self.move(row+2, col+1, move_number+1)
        self.move(row+1, col-2, move_number+1)
        self.move(row+2, col-1, move_number+1)
        self.move(row-1, col-2, move_number+1)
        self.move(row-2, col-1, move_number+1)
        self.move(row-1, col+2, move_number+1)
        self.move(row-2, col+1, move_number+1)
        self.board[row][col] = 0
        return self.solved

    def run(self, row, col):
        return self.move(row, col, 1)


def display_square(square):
    return "{0}{1}".format(chr(65+square[0]), square[1]+1)


if __name__ == "__main__":

    parser = argparse.ArgumentParser()
    parser.add_argument('size', default=7, type=int)
    parser.add_argument('start_col', default=0, type=int)
    parser.add_argument('start_row', default=0, type=int)
    args = parser.parse_args()

    start_pos = display_square((args.start_col, args.start_row))
    print("Starting solver from pos {0}".format(start_pos))
    print("Initializing board. Size={0}".format(args.size))
    solver = KnightSolver(args.size)

    t0 = clock()
    found = solver.run(args.start_col, args.start_row)
    t = clock() - t0
    if found:
        print("Found solution in {0} sec:".format(t))
        for square in solver.path:
            print("{0}{1}".format(chr(65+square[0]), square[1]+1))
    else:
        print("No solution found")
